app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageExpiration.kt

```diff
@@ -0,0 +1,266 @@
+                    }
+
+                    timeLeft <= 24.hours && timeLeft > 1.hours -> {
+                        val timeLeftTillWholeHour = (timeLeft.inWholeSeconds % 1.hours.inWholeSeconds).seconds
+                        if (timeLeftTillWholeHour == ZERO) {
+                            1.hours
+                        } else {
+                            timeLeftTillWholeHour
+                        }
+                    }
+
+                    timeLeft <= 1.hours && timeLeft > 1.minutes -> {
+                        val timeLeftTillWholeMinute = (timeLeft.inWholeSeconds % 1.minutes.inWholeSeconds).seconds
+                        if (timeLeftTillWholeMinute == ZERO) {
+                            1.minutes
+                        } else {
+                            timeLeftTillWholeMinute
+                        }
+                    }
+
+                    timeLeft <= 1.minutes -> {
+                        1.seconds
+                    }
+
+                    else -> throw IllegalStateException("Not possible state for the interval")
+                }
+
+                return timeLeftUpdateInterval
+            }
+
+            fun decreaseTimeLeft(interval: Duration) {
+                if (timeLeft.inWholeSeconds != 0L) timeLeft -= interval
+            }
+
+            fun alphaBackgroundColor(): Float {
+                val totalTimeLeftRatio = timeLeft / expireAfter
+
+                return if (totalTimeLeftRatio >= TIME_LEFT_RATIO_BOUNDARY_FOR_ALMOST_TIME_LEFT_ELAPSED_ALPHA) {
+                    PAST_RATIO_BOUNDARY_FOR_ALMOST_TIME_LEFT_ALPHA_VALUE
+                } else {
+                    BEFORE_RATIO_BOUNDARY_FOR_ALMOST_TIME_LEFT_ALPHA_VALUE
+                }
+            }
+        }
+
+        object NotExpirable : SelfDeletionTimerState()
+    }
+}
+
+@Composable
+fun startDeletionTimer(
+    message: UIMessage,
+    expirableTimer: SelfDeletionTimer.SelfDeletionTimerState.Expirable,
+    onStartMessageSelfDeletion: (UIMessage) -> Unit
+) {
+    message.messageContent?.let {
+        when (val messageContent = message.messageContent) {
+            is UIMessageContent.AssetMessage -> startAssetDeletion(
+                expirableTimer = expirableTimer,
+                onSelfDeletingMessageRead = { onStartMessageSelfDeletion(message) },
+                downloadStatus = messageContent.downloadStatus
+            )
+
+            is UIMessageContent.AudioAssetMessage -> startAssetDeletion(
+                expirableTimer = expirableTimer,
+                onSelfDeletingMessageRead = { onStartMessageSelfDeletion(message) },
+                downloadStatus = messageContent.downloadStatus
+            )
+
+            is UIMessageContent.ImageMessage -> startAssetDeletion(
+                expirableTimer = expirableTimer,
+                onSelfDeletingMessageRead = { onStartMessageSelfDeletion(message) },
+                downloadStatus = messageContent.downloadStatus
+            )
```

Comment:
Can we just group all of these together?
```suggestion
            is UIMessageContent.AssetMessage, is UIMessageContent.ImageMessage, is UIMessageContent.AudioAssetMessage -> startAssetDeletion(
                expirableTimer = expirableTimer,
                onSelfDeletingMessageRead = { onStartMessageSelfDeletion(message) },
                downloadStatus = messageContent.downloadStatus
            )
```

Reply:
not possible unfortunetly :S needs a refactor in order to do this, I am going to leave as it is now, from what I remember Kubaz is going to refactor UiMessage ? If he does not lniclude this in his part I am going to submit a PR later on for this